### PR TITLE
[SSO] spar cleanup

### PR DESCRIPTION
**Pull request by fisx**
- separate config and idpconfig
- remove application logic from interface for storing requests / assertions.
- connect C* tests to C* directly, without going via internal api end-points.
- completely rewrite DataSpec.hs, factor out AppSpec.hs.
- add helpers to run Spar actions from inside TestSpar.

see also: https://github.com/wireapp/saml2-web-sso/pull/34

### CODE REVIEW

File: services/spar/package.yaml [hunk 1]
```diff
@@ -35,8 +35,8 @@ dependencies:
   - http-client-tls
   - http-types
   - interpolate
+  - lens
   - metrics-core
-  - microlens
   - microlens-th
```
Comment (neongreen) [inline services/spar/package.yaml]: This should be removed as well

File: services/spar/test-integration/Test/Spar/AppSpec.hs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                                 , URI                         -- ^ location header
+                                 , [(SBS, SBS)]                -- ^ query params
+                                 )
+requestAccessVerdict isGranted mkAuthnReq = do
+  env <- ask
+  let uid     = env ^. teUserId
+      idp     = env ^. teIdP
+      idpid   = idp ^. SAML.idpId
+      tenant  = idp ^. SAML.idpMetadata . SAML.edIssuer
+      subject = SAML.opaqueNameID "blee"
```
Comment (neongreen) [inline services/spar/test-integration/Test/Spar/AppSpec.hs]: Should this be randomized?

File: services/spar/test-integration/Test/Spar/AppSpec.hs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  runSpar $ Spar.insertUser uref uid
+  authnreq :: SAML.FormRedirect SAML.AuthnRequest <- do
+    raw <- mkAuthnReq idpid
+    bdy <- maybe (error "authreq") pure $ responseBody raw
+    either (error . show) pure $ Servant.mimeUnrender (Servant.Proxy @SAML.HTML) bdy
+  spmeta <- getTestSPMetadata
+  authnresp <- do
+    let mk :: SAML.FormRedirect SAML.AuthnRequest -> TestSpar SAML.AuthnResponse
+        mk (SAML.FormRedirect _ req) = do
+          SAML.SignedAuthnResponse (XML.Document _ el _) <- runSimpleSP $ SAML.mkAuthnResponse SAML.sampleIdPPrivkey idp spmeta req True
```
Comment (neongreen) [inline services/spar/test-integration/Test/Spar/AppSpec.hs]: Lines longer than 78 chars break my screen :(

File: services/spar/test-integration/Util.hs [hunk 4]
```diff
@@ -212,6 +226,20 @@ addTeamMember galleyreq tid mem =
                 . lbytes (Aeson.encode mem)
                 )
 
+nextId :: MonadIO m => m (Id a)
```
Comment (neongreen) [inline services/spar/test-integration/Util.hs]: I'd appreciate comments for these two

Comment (neongreen) [reply services/spar/test-integration/Util.hs]: Also, might `randomId` be a better name?

**Comment by neongreen** [comment]
I'm not familiar with most of the code there, so I won't dare press "Approve"

File: services/spar/test-integration/Test/Spar/AppSpec.hs (hunk 3 above)
Comment (fisx) [reply services/spar/test-integration/Test/Spar/AppSpec.hs]:
```suggestion
          SAML.SignedAuthnResponse (XML.Document _ el _) <- runSimpleSP $
            SAML.mkAuthnResponse SAML.sampleIdPPrivkey idp spmeta req True
```

Comment (fisx) [reply services/spar/test-integration/Test/Spar/AppSpec.hs]: I just did this to try out the suggest-code-changes github feature.  Is it ok if I am going to ignore this in the future?  78 columns is not enough!  At all!

File: services/spar/test-integration/Test/Spar/AppSpec.hs (hunk 2 above)
Comment (fisx) [reply services/spar/test-integration/Test/Spar/AppSpec.hs]: yes, thanks!  will push soon.

File: services/spar/test-integration/Test/Spar/AppSpec.hs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            (uid, outcome, loc, qry) <- requestAccessVerdict True mkAuthnReqMobile
+            liftIO $ do
+              Servant.errHTTPCode outcome `shouldBe` 303
+              Servant.errReasonPhrase outcome `shouldBe` "success"
+              Servant.errBody outcome `shouldBe` mempty
+              uriScheme loc `shouldBe` (URI.Scheme "wire")
+              List.lookup "label"  qry `shouldBe` Nothing
+              List.lookup "userid" qry `shouldBe` (Just . cs . show $ uid)
+              List.lookup "cookie" qry `shouldNotBe` Nothing
+              List.lookup "cookie" qry `shouldNotBe` Just "$cookie"
```
[the remark this answers is not in the document]
Comment (neongreen) [reply services/spar/test-integration/Test/Spar/AppSpec.hs]: Well, I see it and I think "so why would it possibly be $cookie?"

**Pull request merged**